drivers/hmc5883l/hmc5883l.c

```diff
@@ -0,0 +1,338 @@
+
+    /* set operation mode */
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_HS, 0);
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_MD, params->op_mode);
+
+    EXEC_RET(_reg_write(dev, HMC5883L_REG_MODE, &mode, 1), res);
+
+    /* wait 6 ms accoring to data sheet */
+    xtimer_usleep(6 * US_PER_MS);
+
+    return res;
+}
+
+int hmc5883l_data_ready(const hmc5883l_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    int res = HMC5883L_OK;
+
+    uint8_t reg;
+
+    EXEC_RET(_reg_read(dev, HMC5883L_REG_STATUS, &reg, 1), res);
+    return _get_reg_bit(reg, HMC5883L_REG_STATUS_RDY) ? HMC5883L_OK
+                                                      : HMC5883L_ERROR_NO_DATA;
+}
+
+/*
+ * scale factors for conversion of raw sensor data to degree for possible
+ * sensitivities according to mechanical characteristics in datasheet
+ */
+static const uint32_t HMC5883L_RES[] = {
+     730,      /* uG/LSb for HMC5883L_GAIN_1370 with range +-0.88 Gs */
+     920,      /* uG/LSb for HMC5883L_GAIN_1090 with range  +-1.3 Gs */
+    1220,      /* uG/LSb for HMC5883L_GAIN_820  with range  +-1.9 Gs */
+    1520,      /* uG/LSb for HMC5883L_GAIN_660  with range  +-2.5 Gs */
+    2270,      /* uG/LSb for HMC5883L_GAIN_440  with range  +-4.0 Gs */
+    2560,      /* uG/LSb for HMC5883L_GAIN_390  with range  +-4.7 Gs */
+    3030,      /* uG/LSb for HMC5883L_GAIN_330  with range  +-5.6 Gs */
+    4350,      /* uG/LSb for HMC5883L_GAIN_230  with range  +-8.1 Gs */
+};
+
+int hmc5883l_read(const hmc5883l_t *dev, hmc5883l_data_t *data)
+
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(data != NULL);
+    DEBUG_DEV("data=%p", dev, data);
+
+    int res = HMC5883L_OK;
+
+    hmc5883l_raw_data_t raw;
+
+    EXEC_RET(hmc5883l_read_raw (dev, &raw), res);
+
+    data->x = ((uint32_t)raw.x * HMC5883L_RES[dev->params.gain]);
+    data->y = ((uint32_t)raw.y * HMC5883L_RES[dev->params.gain]);
+    data->z = ((uint32_t)raw.z * HMC5883L_RES[dev->params.gain]);
+
+    return res;
+}
+
+int hmc5883l_read_raw(const hmc5883l_t *dev, hmc5883l_raw_data_t *raw)
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(raw != NULL);
+    DEBUG_DEV("raw=%p", dev, raw);
+
+    int res = HMC5883L_OK;
+
+    uint8_t data[6];
+
+    /* read raw data sample */
+    EXEC_RET_CODE(_reg_read(dev, HMC5883L_REG_OUT_X_MSB, data, 6),
+                  res, -HMC5883L_ERROR_RAW_DATA);
+
+    /* data MSB @ lower address */
+    raw->x = (data[0] << 8) | data[1];
+    raw->y = (data[2] << 8) | data[3];
+    raw->z = (data[4] << 8) | data[5];
+
+    return res;
+}
+
+int hmc5883l_power_down (hmc5883l_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    uint8_t mode = 0;
+
+    /* set operation mode to Idle mode with only 5 uA current */
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_HS, 0);
+    _set_reg_bit(&mode, HMC5883L_REG_MODE_MD, HMC5883L_OP_MODE_IDLE);
+
+    return _reg_write(dev, HMC5883L_REG_MODE, &mode, 1);
```

Comment:
This funtions could boil down to a one-liner (+ assertion):
```c
return _reg_write(dev->dev, dev->addr, HMC5883L_REG_MODE, HMC5883L_REG_MODE_MD);
```

Reply:
The question is whether we should use `assert` on an I2C error which can always happen from time to time. From my point of view, `assert` should only be used in case of configuration issues, but not for runtime errors.

Reply:
No, I don't meant to assert the i2c errors, that would be bad practice, indeed! I was rather referring to assert `dev != NULL` and then calling the `_reg_write()` func afterwards. So all I2C error are evaluated inside the `_reg_write()` function during runtime, right?!